src/coordinator.rs

```diff
@@ -78,29 +86,61 @@ impl<'job> Coordinator<'job> {
             )
             .context("could not run job")?;
 
-        for (blocked, blockers) in self.blocked.iter_mut() {
-            if blockers.remove(&next) && blockers.is_empty() {
-                self.ready.push(*blocked);
+        // Now that we're done running the job, we update our bookkeeping to
+        // figure out what running that job just unblocked.
+        //
+        // As an implementation note, this will probably end up in a separate
+        // function once we're running tasks in parallel!
+        let mut newly_unblocked = Vec::default(); // avoiding mutating both fields of self in the loop below
+
+        self.blocked.retain(|blocked, blockers| {
+            let removed = blockers.remove(&next);
+            if !removed {
+                return false;
+            }
 
-                // TODO: it would be more performant to remove the
-                // newly-unblocked item from self.blocked, but there's
-                // already a mutable borrow. Possibly rearrange the code
-                // to do some mutable filtering thing.
+            let no_blockers_remaining = blockers.is_empty();
+            if no_blockers_remaining {
+                newly_unblocked.push(*blocked)
             }
-        }
+            !no_blockers_remaining
+        });
+        self.ready.append(&mut newly_unblocked);
```

Comment:
I don't know this api, but append taking a mutable reference just looks wrong. Maybe some sort of api mistake or other strange design?

Reply:
I agree that it looks very strange. Apparently the borrow type needs to match, and since `self.ready` is mutable at this point the reference to `newly_unblocked` needs to be, too. So if you do:

| Code | Result |
|-|-|
| `self.ready.append(newly_unblocked)` | "expected `&mut Vec<Id>`, found struct `Vec`" |
| `self.ready.append(&newly_unblocked)` | "types differ in mutability" |
| `self.ready.append(&mut newly_unblocked)` | works |

If you have any ideas for how to make this look less strange, I'm all ears!

Reply:
I think that extend should do what you want here and read a tad better: `self.ready.extend(newly_unblocked);`

Reply:
Oh fantastic. Thank you!